test/tests/functional/pbs_ralter.py

```diff
@@ -1516,3 +1550,254 @@ def test_failed_ralter(self):
         self.alter_a_reservation(rid1, start1, end1, shift=300,
                                  alter_e=True, whichMessage=3)
         self.server.expect(JOB, {'job_state': 'Q'}, id=jid)
+
+    def test_adv_resv_duration_before_start(self):
+        """
+        Test duration of reservation can be changed. In this case end
+        time changes, and start time remains the same.
+        """
+
+        offset = 20
+        duration = 20
+        shift = 10
+        rid, start, end = self.submit_and_confirm_reservation(offset, duration)
+
+        self.alter_a_reservation(rid, start, end, a_duration=30,
+                                 check_log=False)
+
+        t_duration, t_start, t_end = self.get_resv_time_info(rid)
+
+        self.assertEqual(t_end, end+10)
+        self.assertEqual(t_start, start)
+        self.assertEqual(t_duration, 30)
+
+        # Submit a job to the reservation and change its duration.
+        self.submit_job_to_resv(rid)
+
+        temp_end = t_end
+        temp_start = t_start
+
+        self.alter_a_reservation(rid, temp_start, temp_end, sequence=2,
+                                 a_duration=40, check_log=False)
+
+        t_duration, t_start, t_end = self.get_resv_time_info(rid)
+        self.assertEqual(t_end, temp_end+10)
+        self.assertEqual(t_start, temp_start)
+        self.assertEqual(t_duration, 40)
+
+    def test_adv_resv_dur_and_endtime_before_start(self):
+        """
+        Test that duration and end time of reservation can be changed together.
+        In this case the start time of reservation may also change.
+        """
+
+        offset = 20
+        duration = 20
+        shift = 10
+        rid, start, end = self.submit_and_confirm_reservation(offset, duration)
+
+        self.alter_a_reservation(rid, start, end, shift=shift,
+                                 alter_e=True, a_duration=40, check_log=False)
+
+        t_duration, t_start, t_end = self.get_resv_time_info(rid)
+
+        self.assertEqual(t_end, end+10)
+        self.assertEqual(t_start, start-10)
+        self.assertEqual(t_duration, 40)
+
+        # Submit a job to the reservation and change its start time.
+        self.submit_job_to_resv(rid)
+        temp_end = t_end
+        temp_start = t_start
+
+        self.alter_a_reservation(rid, temp_start, temp_end, shift=shift,
+                                 alter_e=True, sequence=2,
+                                 a_duration=50, check_log=False)
```

Comment:
Could you modify alter_a_reservation()?  It takes options that are ignored and not needed.  Here you are passing the start and end time, but end time is being altered.  It isn't necessary.  Maybe take default values for start, end, shift, and duration and check which ones are required based on alter_s and alter_e.

Reply:
Left this as is. This function returns the new start and new end, it also to checks start time and end time in log files in all non duration tests. So changing this we will have to change some tests too or calculate new start / endtime in the function. So I think fixing this comment might be an overkill for making test code simpler in a couple of tests.